modules/fedimint-mint-client/src/issuance.rs

```diff
@@ -0,0 +1,372 @@
+///     Outcome -- invalid blind signatures  --> Failed
+///     end
+/// ```
+#[derive(Debug, Clone, Eq, PartialEq, Decodable, Encodable)]
+pub enum MintIssuanceStates {
+    /// Issuance request was created, we are waiting for blind signatures
+    Created(MintIssuanceStatesCreated),
+    /// The transaction containing the issuance was rejected, we can stop
+    /// looking for preimages
+    Aborted(MintIssuanceStatesAborted),
+    /// The transaction containing the issuance was accepted but an unexpected
+    /// error occurred, this should never happen with a honest federation and
+    /// bug-free code.
+    Failed(MintIssuanceStatesFailed),
+    /// The issuance was completed successfully and the e-cash notes added to
+    /// our wallet
+    Succeeded(MintIssuanceStatesSucceeded),
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Decodable, Encodable)]
+pub struct MintIssuanceCommon {
+    pub(crate) operation_id: OperationId,
+    pub(crate) out_point: OutPoint,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Decodable, Encodable)]
+pub struct MintIssuanceStateMachine {
+    pub(crate) common: MintIssuanceCommon,
+    pub(crate) state: MintIssuanceStates,
+}
+
+impl State for MintIssuanceStateMachine {
+    type ModuleContext = MintClientContext;
+    type GlobalContext = DynGlobalClientContext;
+
+    fn transitions(
+        &self,
+        context: &Self::ModuleContext,
+        global_context: &Self::GlobalContext,
+    ) -> Vec<StateTransition<Self>> {
+        match &self.state {
+            MintIssuanceStates::Created(created) => {
+                created.transitions(context, global_context, self.common)
+            }
+            MintIssuanceStates::Aborted(_) => {
+                vec![]
+            }
+            MintIssuanceStates::Failed(_) => {
+                vec![]
+            }
+            MintIssuanceStates::Succeeded(_) => {
+                vec![]
+            }
+        }
+    }
+
+    fn operation_id(&self) -> OperationId {
+        self.common.operation_id
+    }
+}
+
+/// See [`MintIssuanceStates`]
+#[derive(Debug, Clone, Eq, PartialEq, Decodable, Encodable)]
+pub struct MintIssuanceStatesCreated {
+    pub(crate) note_issuance: NoteIssuanceRequests,
+}
+
+impl MintIssuanceStatesCreated {
+    fn transitions(
+        &self,
+        // TODO: make cheaper to clone (Arc?)
+        context: &MintClientContext,
+        global_context: &DynGlobalClientContext,
+        common: MintIssuanceCommon,
+    ) -> Vec<StateTransition<MintIssuanceStateMachine>> {
+        let mint_keys = context.mint_keys.clone();
+        vec![
+            // Check if transaction was rejected
+            StateTransition::new(
+                Self::trigger_tx_rejected(global_context.clone(), common),
+                |_dbtx, (), state| Box::pin(Self::transition_tx_rejected(state)),
+            ),
+            // Check for output outcome
+            StateTransition::new(
+                Self::trigger_outcome_ready(
```

Comment:
Note to self: "trigger_" sounds imperative, while it looks like this code is about awaiting something.